tc/core/tensor_api_rfc.h

```diff
@@ -0,0 +1,163 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <iostream>
+#include <string>
+#include <vector>
+
+#include <dlpack/dlpack.h>
+
+#include "tc/proto/compcache.pb.h"
+
+/**
+ * Various tensor utilities used in the Tensor Comprehensions compiler.
+ * At a high-level we use the DLTensor type from dlpack and we add a
+ * DLConstTensor to propagate const-correctness.
+ * DLTensor is used as an ML-framework and backend-agnostic representation to
+ * interact with user-level tensors.
+ * DLTensorUPtr and DLConstTensorUPtr are used as metadata-owning pointers.
+ * The underlying concrete device pointer is borrowed when calling
+ * ExecutionEngine::compile and ExecutionEngine::run.
+ *
+ * Internally, TC uses TensorInfo for compiling and interacting with the
+ * compilation caches. TensorInfo is backed by protobuf and stored directly in
+ * the caches.
+ */
+
+/**
+ * DLPack is missing a ConstTensor. Add it here for now.
+ * In the future, this would be contributed back to dlpack/include/dlpack.h.
+ */
+typedef struct {
```

Comment:
if we fork this, why not make a better format from the get go, what is the value in using a few dozen of lines in DLPack? Formats are useful for serialization, but here everything is happening in memory.

Storing int64_t*s for the meta info causes all these custom deleters. If you use vectors instead, it would be easier to use. What do you think?

Reply:
> all these custom deleters

do you mean the 10 lines of code for ```struct DLTensorDeleter ```?

We can certainly abandon DLPack for a small pure-C++ abstraction. But the rationale here was that a bunch of external people agreed on using DLPack as the simple abstraction to interoperate and people already implement conversions from/to dlpack that we would have to reimplement for various frameworks. 
This is why we started using DLPack ourselves.

I'd like to keep this consideration separate from the API for now if you don't mind, I'm happy to revisit after the refactoring is done.